src/box/memtx_tx.c

```diff
@@ -1644,6 +1797,85 @@ memtx_tx_track_point_slow(struct txn *txn, struct space *space,
 	return point_hole_storage_new(space->index[index], key, key_len, txn);
 }
 
+static struct gap_item *
+memtx_tx_gap_item_new(struct txn *txn, enum iterator_type type,
+		      const char *key, uint32_t part_count)
+{
+	struct gap_item *item = (struct gap_item *)
+		mempool_alloc(&txm.gap_item_mempoool);
+	if (item == NULL) {
+		diag_set(OutOfMemory, sizeof(*item), "mempool_alloc", "gap");
+		return NULL;
+	}
+
+	item->txn = txn;
+	item->type = type;
+	item->part_count = part_count;
+	const char *tmp = key;
+	for (uint32_t i = 0; i < part_count; i++)
+		mp_next(&tmp);
+	item->key_len = tmp - key;
+	if (part_count == 0) {
+		item->key = NULL;
+	} else if (item->key_len <= sizeof(item->short_key)) {
+		item->key = item->short_key;
+	} else {
+		item->key = (char *)region_alloc(&txn->region, item->key_len);
+		if (item->key == NULL) {
+			mempool_free(&txm.gap_item_mempoool, item);
+			diag_set(OutOfMemory, item->key_len, "tx region",
+				 "point key");
+			return NULL;
+		}
+	}
+	memcpy((char *)item->key, key, item->key_len);
+	rlist_add(&txn->gap_list, &item->in_gap_list);
+	return item;
+}
+
+/**
+ * Record in TX manager that a transaction @a txn have read nothing
+ * from @a space and @ a index with @ key, somewhere from interval between
+ * some unknown predecessor and @a successor.
+ * This function must be used for ordered indexes, such as TREE, for queries
+ * when interation type is not EQ or when the key is not full (otherwise
+ * it's faster to use memtx_tx_track_point).
+ * @return 0 on success, -1 on memory error.
+ */
+int
+memtx_tx_track_gap_slow(struct txn *txn, struct space *space, uint32_t index,
+			struct tuple *successor, enum iterator_type type,
+			const char *key, uint32_t part_count)
+{
+	if (txn->status != TXN_INPROGRESS)
```

Comment:
I'd move this check to memtx_tx_track_gap() as well

Reply:
In `memtx_tx.h`? I thought about that, but this add a dependency from `txn.h`. Are you sure that it is worth it?